extension/golang/environment.go

```diff
@@ -0,0 +1,186 @@
+	"strings"
+	"time"
+
+	"github.com/cloudwan/gohan/db"
+	"github.com/cloudwan/gohan/server/middleware"
+	"github.com/cloudwan/gohan/sync"
+
+	"plugin"
+
+	ext "github.com/cloudwan/gohan/extension"
+	"github.com/cloudwan/gohan/extension/goext"
+	l "github.com/cloudwan/gohan/log"
+	"github.com/cloudwan/gohan/schema"
+)
+
+var log = l.NewLogger()
+
+// Environment golang based rawEnvironment for gohan extension
+type Environment struct {
+	Name       string
+	DataStore  db.DB
+	timeLimit  time.Duration
+	timeLimits []*schema.EventTimeLimit
+	Identity   middleware.IdentityService
+	Sync       sync.Sync
+
+	// golang extension environment
+	extEnvironment goext.Environment
+}
+
+// NewEnvironment create new gohan extension rawEnvironment based on context
+func NewEnvironment(name string, dataStore db.DB, identity middleware.IdentityService, sync sync.Sync) *Environment {
+	env := &Environment{
+		Name:      name,
+		DataStore: dataStore,
+		Identity:  identity,
+		Sync:      sync,
+	}
+	env.SetUp()
+	return env
+}
+
+// SetUp initialize rawEnvironment
+func (environment *Environment) SetUp() {
+	environment.extEnvironment.Handlers = goext.EventPriorityHandlerList{}
+
+	environment.extEnvironment.Core = bindCore(environment)
+	environment.extEnvironment.Logger = bindLogger(environment)
+	environment.extEnvironment.Schemas = bindSchemas(environment)
+}
+
+//Load loads script for rawEnvironment
+func (environment *Environment) Load(source, code string) error {
+	log.Debug("Loading golang extension: %s", source)
+
+	if filepath.Ext(source) != ".so" {
+		return fmt.Errorf("Golang extensions source code must be a *.so file, source: %s", source)
+	}
+
+	p, err := plugin.Open(source)
+
+	if err != nil {
+		return fmt.Errorf("Failed to load golang extension: %s", err)
+	}
+
+	ifn, err := p.Lookup("Init")
+
+	if err != nil {
+		return fmt.Errorf("Golang extension does not export extensionInit: %s", err)
+	}
+
+	log.Debug("Init golang extension: %s", source)
+
+	err = ifn.(func(*goext.Environment) error)(&environment.extEnvironment)
+
+	if err != nil {
+		log.Error("Failed to initialize golang extension: %s; error: %s", source, err)
+		return err
+	}
+
+	log.Debug("Golang extension initialized: %s", source)
+
+	return nil
+}
+
+//LoadExtensionsForPath for returns extensions for specific path
+func (environment *Environment) LoadExtensionsForPath(extensions []*schema.Extension, timeLimit time.Duration, timeLimits []*schema.PathEventTimeLimit, path string) error {
+	for _, extension := range extensions {
+		if extension.Match(path) {
+			code := extension.Code
+			if extension.CodeType != "go" {
+				continue
+			}
+			if extension.URL == "" {
+				log.Warning("Found golang extension without plugin - ignored")
+				continue
+			}
+			url := strings.TrimPrefix(extension.URL, "file://")
+			err := environment.Load(url, code)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	// setup time limits for matching extensions
+	environment.timeLimit = timeLimit
```

Comment:
timeLimits unused? by the way, how are we going to handle timeouts? golangs (not Gohans) contexts?